core/store/src/db/bytes.rs

```diff
@@ -0,0 +1,131 @@
+use std::sync::Arc;
+
+use super::refcount;
+
+/// Data returned from the database.
+///
+/// Abstraction layer for data returned by [`Database::get_raw_bytes`] and
+/// [`Database::get_with_rc_stripped`] methods.  Operating on the value as
+/// a slice is free while converting it to a vector on an arc may requires an
+/// allocation and memory copy.
+pub struct DBBytes<'a>(Inner<'a>);
+
+enum Inner<'a> {
+    /// Data held as a vector.
+    Vec(Vec<u8>),
+
+    /// Data held as RocksDB-specific pinnable slice.
+    Rocks {
+        /// Pointer at the bytes.
+        ///
+        /// The data is held by `db_slice` and as such this field must not
+        /// outlive `db_slice`.  Nor can `db_slice` be modified.
+        ///
+        /// We keep a separate pointer because we want to be able to modify
+        /// truncate the value without having to allocate new buffers.
+        data: *const [u8],
+
+        /// This is what owns the data.
+        ///
+        /// This is never modified which is why we can have `data` field
+        /// pointing at the buffer this object holds.
```

Comment:
This is a bit impercise, as what matters is that data has a stable deref, not that it is not modified. Ie, if this was `db_slice: [u8; 42]`, the `data` pointer could've become invalidated by moves even without direct modifications to db_slice itself.